client/pdf_generator.py

```diff
@@ -0,0 +1,175 @@
+			display_name='* Template File'
+		),
+		plugins.ClientOptionString(
+			'logo',
+			'Image to include into the pdf',
+			display_name="Logo Image"
+		),
+		plugins.ClientOptionString(
+			'link_text',
+			'Text for inserted link',
+			default='Click here to accept',
+			display_name="Link Text"
+		)
+	]
+	req_min_version = '1.7.0b1'
+	version = '1.1'
+
+	def initialize(self):
+		mailer_tab = self.application.main_tabs['mailer']
+		self.text_insert = mailer_tab.tabs['send_messages'].text_insert
+		self.add_menu_item('Tools > Create PDF Preview', self.make_preview)
+		self.signal_connect('send-precheck', self.signal_send_precheck, gobject=mailer_tab)
+		self.signal_connect('send-target', self.signal_send_target, gobject=mailer_tab)
+		self.signal_connect('send-finished', self.signal_send_finished, gobject=mailer_tab)
+		return True
+
+	def signal_send_precheck(self, _):
+		if not any((self.config['template_file'], self.config['output_pdf'], self.config['link_text'])):
+			self.logger.debug('skipping exporting any attachments due to lack of information provided to run')
+			return True
+
+		if not os.path.isfile(self.config['template_file']):
+			self.logger.error('no pdf template file found')
+			return
+		else:
+			self.logger.debug('pdf template file found, generating attachment')
+			return True
+
+		if self.config['logo']:
+			if not os.path.isfile(self.config['logo']):
+				self.logger.error('logo image specified but cannot be opened')
+				return
+
+	def make_preview(self, _):
+		outfile = self.expand_path(self.config['output_pdf'])
+		pdf_file = SimpleDocTemplate(outfile, pagesize=letter,
+	                        rightMargin=72, leftMargin=72,
+	                        topMargin=72, bottomMargin=18)
+		url = self.application.config['mailer.webserver_url']
+		pdf = self.get_template(url)
+		pdf_file.multiBuild(pdf)
+		self.logger.info('pdf preview created. Check ' + self.config['output_pdf'])
+
+	def signal_send_target(self, _, target):
+		outfile = self.expand_path(self.config['output_pdf'])
+		pdf_file = SimpleDocTemplate(outfile, pagesize=letter,
+	                    	rightMargin=72, leftMargin=72,
+	                        topMargin=72, bottomMargin=18)
+		url = self.application.config['mailer.webserver_url'] + '?uid=' + target.uid
+		pdf = self.get_template(url)
+		try:
+			pdf_file.multiBuild(pdf)
+			self.attach_pdf(outfile)
+			self.logger.info('pdf attachement made linking with uid: ' + target.uid)
+		except Exception as err:
+			self.logger.error('pdf could not be oppened: ', err)
+
+
+	def get_template(self, url):
+		logo = self.config['logo']
+		formatted_time = time.ctime()
+		company = self.application.config['mailer.company_name']
+		sender = self.application.config['mailer.source_email_alias']
+
+		Story = []
+		click_me = self.config['link_text']
+		link = '<font color=blue><link href="' + str(url) + '">' + click_me + '</link></font>'
+		if self.config['logo']:
+			im = Image(logo, 2*inch, 1*inch)
```

Comment:
Operators should have a single space on either side of them, like: `2 * inch, 1 * inch`. Also 1x anything is anything so it shouldn't be necessary.